python/pyspark/sql/session.py

```diff
@@ -492,28 +539,40 @@ def _inferSchema(self, rdd, samplingRatio=None, names=None):
                 prefer_timestamp_ntz=prefer_timestamp_ntz)).reduce(_merge_type)
         return schema
 
-    def _createFromRDD(self, rdd, schema, samplingRatio):
+    def _createFromRDD(
+        self,
+        rdd: "RDD[Union[DateTimeLiteral, LiteralType, DecimalLiteral, RowLike]]",
+        schema: Optional[Union[DataType, List[str]]],
+        samplingRatio: Optional[float],
+    ) -> Tuple["RDD[Tuple]", StructType]:
```

Comment:
Following the notes from the above, this could be overloaded to distinguish between cases were we can and cannot infer schema. Might be an overkill, though.

Just a heads-up ‒ I've encountered some problems related to these specific `Unions` while working on SPARK-36894. This surface only with the `self` type (which is, ironically, not validated) and I am thinking about introducing some `TypeVars` (a more precise choice anyway) as a fix.

Reply:
Sure, I'll wait for it and use the `TypeVar` here and the above.

Reply:
> Sure, I'll wait for it and use the TypeVar here and the above.

Oh, I didn't mean that. If any changes are needed later, I'll handle it.